### PR TITLE
Refactor user invitation

### PR DESCRIPTION
**Pull request by allenwq**
- Directly store the name and email in invitation record
   It decouples the invitation and course_user/email in schema, and https://github.com/Coursemology/coursemology2/issues/735 gets properly fixed and course user does not get created when inviting
- Moves the all invitation logic from `course_users_controller` to `course_user_invitations_controller`

Future TODO:
- Remove the workflow state of course_users, it's a mental burden to always remember `with_approved_state` in everywhere of the code.
- Guide the user to create the account after user clicks the confirmation link

Some thoughts:
When refactoring, my time spent on fixing the failing tests are much longer than the actual implementing. 
I think we should focus more on high level feature specs instead of the unit tests. And even for unit tests, it's better to test the public API than the implementation details (private methods). Because the implementation will change when refactoring, but the public interfaces will likely remain unchanged.

### CODE REVIEW

**Comment by weiqingtoh** [comment]
> When refactoring, my time spent on fixing the failing tests are much longer than the actual implementing.
>
> I think we should focus more on high level feature specs instead of the unit tests. And even for unit tests, it's better to test the public API than the implementation details (private methods). Because the implementation will change when refactoring, but the public interfaces will likely remain unchanged.

👍 I'm trying not to test private methods now, but in general I would add specs for the more important ones, in case we reuse them (or refactor them). We can go with this approach?

File: app/views/course/user_invitations/index.html.slim [hunk 1]
```diff
@@ -0,0 +1,23 @@
+- add_breadcrumb :index
+= page_header
+
+= render partial: 'course/users/tabs'
+
+- accepted_invitations = @invitations.each.select(&:confirmed?)
+- pending_invitations = @invitations.select { |invitation| !invitation.confirmed? }
```
Comment (allenwq) [inline app/views/course/user_invitations/index.html.slim]: 👍 You reminded me that there's `partition`.  I used `select` before cause it has better performance than array subtraction ( O(n) vs (O(m*n) )

File: config/routes.rb [hunk 2]
```diff
@@ -199,6 +199,8 @@
       resources :levels, except: [:show, :edit, :update]
       resource :duplication, only: [:show, :create]
 
+      resources :user_invitations, except: [:edit, :update]
```
Comment (allenwq) [inline config/routes.rb]: Thanks, there's no `show` in that controller.

**Comment by allenwq** [comment]
@weiqingtoh Okay. There are some cases which are hard to test using high level specs, or the logic is important and complex, think it's good to have unit test to test them in detail at those times.

**Comment by allenwq** [comment]
@weiqingtoh updated, and added a new commit to fix some UX issue.

**Comment by weiqingtoh** [comment]
Just a random question, if:
- User `A` has emails `a@a.com` & `b@a.com`
- We have 2 invitations - `a@a.com`and `b@a.com`. 

If we accept `a@a.com`, how can we deal with the other outstanding invitation? 
Will it be affected by whether we allow for email changes to invitations?

**Comment by weiqingtoh** [comment]
My guess is that accepting `a@a.com` is smooth and they will create a `course_user` with the correct attributes. If we try to accept `b@a.com`, and create a `course_user`, there will be a `unique_index` error thrown, so it will not be possible to accept the invitation?

**Comment by allenwq** [comment]
They are considered different invitations to different emails, each invitation has its own key, it depends on which key the user confirms first.

If the user has 2 different accounts, then both of them can be enrolled into the course.
If two emails are under the same user and the user manages to input the invitation key twice. yes, there will be a not unique error from course_user.

Normally, he cannot do that because after he inputed the key, he's already been enrolled and there's no UI for inputting again. Unless he open two browser tabs and input seprately  ?

**Pull request merged**